pageserver/src/walredo.rs

```diff
@@ -226,71 +233,68 @@ impl PostgresRedoManager {
     fn apply_batch_postgres(
         &self,
         key: Key,
-        lsn: Lsn,
+        _lsn: Lsn,
         base_img: Option<Bytes>,
         records: &[(Lsn, NeonWalRecord)],
-        wal_redo_timeout: Duration,
+        _wal_redo_timeout: Duration,
         pg_version: u32,
     ) -> Result<Bytes, WalRedoError> {
         let (rel, blknum) = key_to_rel_block(key).or(Err(WalRedoError::InvalidRecord))?;
 
         let start_time = Instant::now();
+        {
+            let mut process_guard = self.process.lock().unwrap();
 
-        let mut process_guard = self.process.lock().unwrap();
-        let lock_time = Instant::now();
-
-        // launch the WAL redo process on first use
-        if process_guard.is_none() {
-            let p = PostgresRedoProcess::launch(self.conf, self.tenant_id, pg_version)?;
-            *process_guard = Some(p);
+            //
+            // Launch the WAL redo process on first use
+            //
+            // TODO-1: it requires locking mutex oin each redo request. It will be better to launch process at
+            // PostgresRedoManage::new but then we should find a way to prevent laucnhing walredo process on
+            // bootsrap incation of pageserver, otherwise two initdb may confict with ah other.
+            //
+            // TODO-2: we are not reading stderr of walredo process right now and so not handling waredo errors.
```

Comment:
Could do the assumption that whatever has been logged to stderr will be logged during startup, or will be logged while processing work item (and the pipe is large enough for it).

What has worked for me in the past is to just have the external process log an initialization complete message `"ready\n"` once all initialization is complete. That should not be logged, but the absence of it in some time or something other should be, and also regarded as a failure.

For the second case, we could just read the stderr after reading the page. On "no issues", there would be many messages but always finally `"done\n"` or something similar, which should not be logged.

Reply:
It is not clear who and when will read this walredo stderr. Launch special tokio task for it? But if the only goal of this task will be to log this errors, then may be it is easier to perform logging at walredo side? Just do not log errors to stderr but save them in some log file (i.e `walredo.log` or logs/pg.log in `wal-redo-datadir`)?
In any case, I consider errors in walredo as something critical and abnormal. Yes, we do not expect safety from redo handles, so i principle compute can send some incorrect walredo requests which can crash walerdo process or let it do some wrong things. We should just enforce that it will not affect other tenants and pageserver will continue to server them.